Review azarashin

azarashin opened the pull request:
とりあえず試しにレビューしてみました。
プルリク受け付けてればよいですが…。

折角なので他のコードも見てみます。
（たまにくらいはレビューしないと感覚が鈍ってヤバイので…）

azarashin committed:
指摘を追加。

azarashin committed:
Reduced nesting.

azarashin committed:
add comment.

azarashin committed:
reviewed then modified.

azarashin committed:
add comment.

azarashin committed:
reviewed then modified.

azarashin committed:
add comment.

azarashin committed:
add comments.

azarashin committed:
review then modified.

azarashin committed:
add comments.

azarashin committed:
reviewed then modified.
(note: It's also better to add more comments and codes for abnormal situation.)

azarashin committed:
add comments.

azarashin committed:
reviewed then modified.

azarashin committed:
add comments.

azarashin committed:
reviewed then modified.

ikeay commented:
ありがとうございます！確認します＼( `･ω･)┐ｼｭﾀｯ

azarashin committed:
add comments.

azarashin committed:
reviewed then modified.

ikeay commented on caesar/caesar.c:
```diff
@@ -7,14 +7,15 @@ int my_atoi(char *str);
         return 1;
     }
 
-    int key = my_atoi(argv[1]);
```
`./caesar 20x` など、このクラスを除くと数字＋文字列の組み合わせでくると受け付けてしまいます。 
上記のような組み合わせが来た場合も、`printf("Usage: ./caesar key\n");` と出力したいです。

ikeay commented on caesar/caesar.c:
```diff
@@ -7,14 +7,15 @@ int my_atoi(char *str);
-    int key = my_atoi(argv[1]);
-    if (key == -1)
+	int key; 
+	int ret = sscanf(argv[1], "%d", &key);
```
ここのロジックわかりやすいですね。C言語のライブラリを全然知らないので参考になります。

azarashin replied:
C99 とANSI C のようにＣ言語にもいくつか種類があるので、
微妙な奴は仕様を確認しておいた方が良いですね。
(今回のatoi とか)

ikeay commented on cash/cash.c:
```diff
@@ -6,6 +6,11 @@ int main(void)
+		// お金を扱う時は、扱える値の範囲に注意です。
+		// 実運用の時は比較的容易にオーバーフローを起こします。
+		// 一部のSQL の中にはmoney型とかいうものもあります。
+		// (固定小数で64bit 表現)
```
ありがとうございます！参考になります。

azarashin replied:
^-^

ikeay commented on credit/credit.c:
```diff
@@ -24,8 +24,8 @@ int main(void)
         {
-            printf("INVALID\n");
-            return 0;
+            printf("INVALID: out of range.\n");
```
これは課題の規定なのですが、 `INVALID` のみ出力するよう指定されています。
https://cs50.harvard.edu/x/2020/psets/1/credit/

ikeay commented on credit/credit.c:
```diff
@@ -74,7 +73,7 @@ int main(void)
-        printf("INVALID\n");
-        return 0;
+        printf("INVALID: bad check sum.\n");
+        return -2;
```
これも課題の規定で 0 を返すように規定されています。

azarashin replied:
この辺もろもろ了解です。
（結構細かいですね…）

ikeay commented on credit/credit.c:
```diff
@@ -24,8 +24,8 @@ int main(void)
-            printf("INVALID\n");
-            return 0;
+            printf("INVALID: out of range.\n");
+            return -1;
```
これは課題の規定で 0 を返すように指定されています。

ikeay commented on credit/credit.c:
```diff
@@ -74,7 +73,7 @@ int main(void)
         }
-        printf("INVALID\n");
-        return 0;
+        printf("INVALID: bad check sum.\n");
```
こちらも `INVALID` のみ出力するよう指定されています。

ikeay commented on credit/credit.c:
```diff
@@ -40,8 +40,7 @@ int main(void)
                 {
-                    sum += _mult / 10;
-                    sum += _mult % 10;
+					sum += _mult - 9;
```
そっか、そうですよねえ、よくよく考えたら…

azarashin committed:
add sample code for pointer.

ikeay commented on mario/less/mario.c:
```diff
@@ -22,6 +22,9 @@ int main(void)
+        } else {
+			printf("height must be 1 to 8. your input is %d\n", number);
+		}
+		return -1;
```
不正な数字が来たら、繰り返し入力させなくてはいけない課題指定になっております。
https://cs50.harvard.edu/x/2020/psets/1/mario/less/

azarashin replied:
・・・単純に、return -1 の場所を間違えました。。。

azarashin replied:
エラーメッセージはとりあえず
コメントアウトだけしておきましょうかね。

azarashin replied:
-1 ではなく、0 返しておきましょうか。
値を返さないで関数を終了するパスがあるのは
まずいので、その点だけ考慮しておく方向で。

azarashin commented:
ポインタ変数の使い方例を追加してみました(pointer.c)。

こいつとにらめっこしてみるとポインタの理解が深まる。かも。

ikeay commented on mario/more/mario.c:
```diff
@@ -10,31 +10,23 @@ int main(void)
+				for (int j = 0; j < i; j ++)
+				{
+					printf("#");
+				}
```
確かにすぎる（後ろの空白も出さなきゃいけないと思ってたときのコードが残りっぱなしになってました）
ご指摘ありがとうございます。

ikeay commented on mario/more/mario.c:
```diff
@@ -10,31 +10,23 @@ int main(void)
     }
-}
\ No newline at end of file
+	return -1; 
```
0は課題的に返してはいけないみたいです。変な入力があったらwhileで続けて入力させる仕様です。

azarashin replied:
main で返却値をreturn せずに関数を抜けるのがまずいので、
そこだけ考慮するようにしておきましょうかね。

ikeay commented on plurality/plurality.c:
```diff
@@ -55,31 +64,20 @@ bool vote(char *name)
+		if(candidates[i].votes == max) {
+			printf("%s\n", candidates[i].name);
+		}
+	}
```
確かにこちらのアルゴリズムのほうがスッキリしますね

azarashin replied:
^-^

ikeay commented on plurality/plurality.c:
```diff
@@ -19,15 +19,24 @@ int candidate_count;
-    for (int i = 0; i < argc - 1; i++)
+	if(argc - 1 > MAX) {
+		fprintf(stderr, "too many parameters. number of parameters must be less than %d\n", MAX); 
+		return -1; 
```
ceaser.c では課題の指定でエラーコード `1` 返しているので、このプロジェクトでは一律エラーコードは `1` としてもよいですか？

azarashin replied:
そうしましょう。

azarashin replied:
ところで、実行時に引数を書かない場合(candidate_count == 0になる場合)、
候補者がいない状態になるので、一応異常ケースを追加しておいた方がよいですかね。
（課題の仕様次第ですが…）

ikeay commented on plurality/plurality.c:
```diff
@@ -19,15 +19,24 @@ int candidate_count;
     int voters_num = get_int("Number of voters: ");
+	if(voters_num <= 0) {
+		fprintf(stderr, "invalid number of voters: %d\n", voters_num); 
+		return -2; 
```
ここもこちらでよいですか？

> ceaser.c では課題の指定でエラーコード `1` 返しているので、このプロジェクトでは一律エラーコードは `1` としてもよいですか？

azarashin replied:
ok

ikeay commented on readability/readability.c:
```diff
@@ -48,4 +44,17 @@ int main(void)
\ No newline at end of file
+}
+
+int ColemanLiauIndex(int letter_count, int word_count, int sentence_count)
```
こちらの関数名ですが、他の課題ではスネークケースになっていいるので、スネークケースでお願いしても良いでしょうか 🙇

ikeay replied:
関数化したほうがスッキリわかりやすいですね。

azarashin replied:
命名規則っぽいのがありましたか。合わせておきましょう。

ikeay commented on substitution/substitution.c:
```diff
@@ -2,63 +2,48 @@
+	for(int i = 0;text[i] != '\0';i++) {
+		int index = (int)text[i]; // 1バイトデータなので、index は0～255 になる。
+		printf("%c", convert_map[index]); 
+	}
```
for文の前後を `printf("ciphertext:");` `printf("\n");` で囲っていただけると 🙇

azarashin replied:
これもspec. に書いてましたね…orz

"Your program must output ciphertext: (without a newline) followed by the plaintext’s corresponding ciphertext, with each alphabetical character in the plaintext substituted for the corresponding character in the ciphertext; non-alphabetical characters should be outputted unchanged.
"

azarashin replied:
余談ですが、今回の修正ででてきたconvert_map のように、１対１対応した変換処理を
配列を使って実現してやると、処理がシンプルになるので処理が高速になります。
但し、変換マップ分のメモリ(256byte)が必要になるので、処理速度とメモリのトレードオフになります。

マイコンプログラムとかの場合だとこの程度のメモリ消費でも致命的になることがあるので注意です。
敢えて少々処理時間がかかっても、処理頻度が低ければメモリ消費を抑える選択をすることもあります。

ikeay replied:
なるほどです

ikeay commented on substitution/substitution.c:
```diff
@@ -2,63 +2,48 @@
-        {
-            if (keys[i] == keys[j] && i != j)
-            {
-                printf("Duplicate characters in key.\n");
```
キーが被った場合、暗号を複合できないので弾かないといけなさそうです。ここを消すと被ってしまいます。

azarashin replied:
Spec. のこの辺の記述ですね。
"If the key is invalid (as by not containing 26 characters, containing any character that is not an alphabetic character, or not containing each letter exactly once), your program should print an error"

確かに、エラーチェックの部分は残しておく必要がありますね。
具体的には、
- keys はアルファベット26文字であること（ぬけがあってはいけない）
- keys にアルファベット以外のものが含まれてはいけない
- keys の各要素が互いに重複しないこと(大文字・小文字が混在していても大丈夫なように工夫する)
でしょうか。。。
大文字・小文字の混在ケースに対してはどこかでケアしておく必要がありますね。

一方で、KeyNums[i] の部分を残しておくのはmust  ではないようですね。

ikeay commented on substitution/substitution.c:
```diff
@@ -2,63 +2,48 @@
-            }
-        }
-    }
+	if(strlen(keys) != 26) {
```
キーはアルファベットの大文字もしくは小文字のみの規定です…！

azarashin replied:
ここはもうちょっと説明が必要そうですね。
（というところでぼちぼち仕事…。また夜くらいに見てみます。）

azarashin replied:
https://cs50.harvard.edu/x/2020/psets/2/substitution/

Specification を見てみると、"The key itself should be case-insensitive, so whether any character in the key is uppercase or lowercase should not affect the behavior of your program." とあるので、大文字と小文字が混在しててもマッピングできるようにしてね、ということですかね（混在してたらエラー表示、とかいう機能は不要っぽい）。

azarashin replied:
このソースの更新は少し細かめに段階刻んでおきましょうかね。。

ikeay commented on substitution/substitution.c:
```diff
@@ -2,63 +2,48 @@
-    }
+	if(strlen(keys) != 26) {
+		printf("Key must contain 26 characters.\n");
+		return -1; 
```
課題の規定なので、エラーは `1` を返していただけると！

ikeay reviewed (commented):
プルリクありがとうございます。

ikeay commented on pointer/pointer.c:
```diff
@@ -0,0 +1,160 @@
+	ex_tolower(); 
+
+	return 0; 
+}
```
わー！ありがとうございます。後ほど確認します 👍

azarashin committed:
restored my_atio.
let me review again later.

azarashin committed:
add comment again.

azarashin committed:
reviewed then modified again.

(this modification will not receive negative value as command line parameter.
if we make it possible also to receive negative value,
this code need to be modified much more.)

azarashin committed:
modified message.

azarashin committed:
modified to follow pset.

azarashin committed:
modified to follow pset.

azarashin committed:
modified to follow pset.

azarashin committed:
modified to follow pset.
Also, I add code for abnormal case.

azarashin committed:
modified message.

azarashin committed:
modified to follow pset.

azarashin committed:
fixed code to follow spec.
"If the key is invalid (as by not containing 26 characters, containing any character that is not an alphabetic character, or not containing each letter exactly once), your program should print an error"

azarashin committed:
fixed code to follow the spec.
"... your program should print an error message of your choice (with printf) and return from main a value of 1 (which tends to signify an error) immediately.".

azarashin committed:
fixed code to follow spec.
"Your program must output ciphertext: (without a newline) followed by the plaintext substituted for the corresponding character in the ciphertext; non-alphabetical characters should be outputted unchanged."

Commit:
fixed style

Commit:
delete comments

Commit:
changed tabs to spaces

Commit:
removed comments and replaced tabs to spaces

Commit:
replaced tabs to spaces

Commit:
replaced tabs to spaces

Commit:
replaced tabs to spaces

Commit:
replaced tabs to spaces and replace camelCase to snake_case

Commit:
replaced tabs to spaces and inserted blanks

Commit:
removed a trash

Commit:
uses

Commit:
changed comments to use English

Commit:
simplify

Commit:
remove pointer/pointer.c

Commit:
deleted NULL definition

azarashin committed:
Merge pull request #1 from ikeay/fixed_style

スタイルを統一する

ikeay commented:
@azarashin よろしければマージします👀

azarashin commented:
確認しました。マージして大丈夫です。
